File: unidbg-android/src/main/java/com/vortexdbg/linux/android/AndroidEmulatorBuilder.java
package com.vortexdbg.linux.android;

import com.vortexdbg.AndroidEmulator;
import com.vortexdbg.EmulatorBuilder;

public class AndroidEmulatorBuilder extends EmulatorBuilder<AndroidEmulator> {

    public static AndroidEmulatorBuilder for32Bit() {
        return new AndroidEmulatorBuilder(false);
    }

    public static AndroidEmulatorBuilder for64Bit() {
        return new AndroidEmulatorBuilder(true);
    }

    protected AndroidEmulatorBuilder(boolean is64Bit) {
        super(is64Bit);
    }

    @Override
    public AndroidEmulator build() {
        return is64Bit ? new AndroidARM64Emulator(processName, rootDir, backendFactories) : new AndroidARMEmulator(processName, rootDir, backendFactories);
    }

}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/linux/android/AndroidEmulatorBuilder.kt
package com.vortexdbg.linux.android

import com.vortexdbg.AndroidEmulator
import com.vortexdbg.EmulatorBuilder

open class AndroidEmulatorBuilder protected constructor(is64Bit: Boolean) : EmulatorBuilder<AndroidEmulator>(is64Bit) {

    override fun build(): AndroidEmulator {
        return if (is64Bit) {
            AndroidARM64Emulator(processName, rootDir, backendFactories)
        } else {
            AndroidARMEmulator(processName, rootDir, backendFactories)
        }
    }

    companion object {
        @JvmStatic
        fun for32Bit(): AndroidEmulatorBuilder {
            return AndroidEmulatorBuilder(false)
        }

        @JvmStatic
        fun for64Bit(): AndroidEmulatorBuilder {
            return AndroidEmulatorBuilder(true)
        }
    }

}
